internal/audit/hub/filter.go

```diff
@@ -0,0 +1,474 @@
+		}
+	}
+
+	tb.buf += string(r)
+
+	return nil
+}
+
+func (tb *tokenBuilder) Flush() error {
+	switch tb.s {
+	case stateClosed, statePlainAccessor:
+	case stateStringOpen:
+		return errors.New("invalid string not closed")
+	default:
+		return fmt.Errorf("flush called in an invalid state: %v", tb.s)
+	}
+
+	if tb.size > 0 {
+		var value any
+		switch tb.t { //nolint:exhaustive
+		case tokenAccessor:
+			value = tb.buf
+		case tokenIndex:
+			idx, err := strconv.Atoi(tb.buf)
+			if err != nil {
+				return err
+			}
+			value = idx
+		}
+
+		t := &Token{
+			typ: tb.t,
+			val: value,
+		}
+
+		if tb.curToken.children == nil {
+			tb.curToken.children = make(map[string]*Token)
+		}
+
+		if cached, ok := tb.curToken.children[t.key()]; ok {
+			tb.curToken = cached
+		} else {
+			tb.curToken.children[t.key()] = t
+			tb.curToken = t
+		}
+	}
+
+	tb.buf = ""
+	tb.t = tokenUnknown
+	tb.size = 0
+
+	return nil
+}
+
+func Tokenize(root *Token, path string) error {
+	curs := 0
+	b := &tokenBuilder{
+		curToken: root,
+	}
+
+	var prev rune
+	for curs < len(path) {
+		r, size := utf8.DecodeRuneInString(path[curs:])
+		curs += size
+
+		// handle and validate token boundaries
+		switch r {
+		case '.':
+			switch {
+			case curs == size:
+				return errors.New("invalid first character: '.'")
+			case curs == len(path):
+				return errors.New("invalid final character: '.'")
+			case prev == '.':
+				return errors.New("invalid empty token")
+			}
+
+			if err := b.Flush(); err != nil {
+				return err
+			}
+		case '[':
+			if err := b.Flush(); err != nil {
+				return err
+			}
+			if err := b.WriteRune(r); err != nil {
+				return err
+			}
+		case ']':
+			if err := b.WriteRune(r); err != nil {
+				return err
+			}
+			if err := b.Flush(); err != nil {
+				return err
+			}
+		default:
+			if err := b.WriteRune(r); err != nil {
+				return err
+			}
+		}
+
+		prev = r
+	}
+
+	if err := b.Flush(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (f *AuditLogFilter) Filter(ingestBatch *logsv1.IngestBatch) error {
+	if f.astRoot == nil {
+		return nil
+	}
+
+	for _, c := range f.astRoot.children {
+		visit(c, ingestBatch.ProtoReflect())
+	}
+
+	return nil
+}
+
+// We support a subset of JSONPath operations, as follows:
+//
+// - dot notation: `foo.bar.baz`
+// - or bracket-notation: `['foo']['bar']['baz]`
+// - or combinations thereof
+//
+// `bar` or `baz` above can be map keys, nested messages or structs.
+//
+// We support list indexing with Ints or wildcards:
+// - foo.bar[0]
+// - foo.bar[*]
+//
+// Wildcards can also operate on member names as a match-all. E.g `foo[*].baz`
+// will match both `baz` values in the pseudo-object below:
+//
+//	{
+//	  'foo': {
+//	    'pow': {
+//	        'baz',
+//	    },
+//	    'bosh': {
+//	        'baz',
+//	    },
+//	  }
+//	}
+func visit(t *Token, m protoreflect.Message) {
+	if value, ok := m.Interface().(*structpb.Value); ok {
+		visitStructpb(t, value)
+		return
+	}
```

Comment:
Just avoiding an allocation.

```suggestion
    if m.Type().Descriptor().FullName() == "google.protobuf.Value" {
		visitStructpb(t, m.Interface().(*structpb.Value))
		return
	}
```